sql/sql_class.h

```diff
@@ -1566,6 +1567,70 @@ typedef struct
   ulonglong start_utime;
 } QUERY_START_TIME_INFO;
 
+/**
+   A single-hash-function bloom filter for approximate accessed page
+   counter.
+*/
+class Bloom_filter {
+ private:
+  /** Bloom filter size */
+  enum { SIZE= 8192 };
+
+  typedef std::bitset<SIZE> Bit_set;
+
+  /** The bit set, which is allocated in a MEM_ROOT */
+  Bit_set *bit_set;
+
+  // Non-copyable
+  Bloom_filter (const Bloom_filter &);
+  Bloom_filter & operator = (const Bloom_filter &);
+
+ public:
+  Bloom_filter() : bit_set(NULL) {}
+
+  ~Bloom_filter()
+  {
+    // Do not delete, just destruct, due to MEM_ROOT allocation
+    bit_set->~bitset();
+  }
+
+  void clear()
+  {
+    // Do not delete, just force new MEM_ROOT allocation on the next use
+    bit_set= NULL;
+  }
+
+  /**
+     Check whether key is maybe a member of a set
+
+     @param[in, out]    mem_root        MEM_ROOT to allocate the bit set in, if
+      not allocated already
+     @param[in]         key             key whose presence to check
+
+     @return if true, the key might be a member of the set. If false, the key
+     is definitely not a member of the set.
+  */
+  bool test_and_set(MEM_ROOT *mem_root, ulong key)
+  {
+    if (!bit_set)
+    {
+      void *bit_set_place= alloc_root(mem_root, sizeof(Bit_set));
+      // FIXME: memory allocation failure is eaten silently. Nonexact stats are
+      // the least of the concerns then.
+      if (!bit_set_place)
+        return false;
+      bit_set= new (bit_set_place) Bit_set();
+    }
+    // Duplicating ut_hash_ulint calculation
+    const ulong pos= (key ^ 1653893711) % SIZE;
```

Comment:
`ut_hash_ulint()` assumes that the second parameter (modulus) is a prime number.
```c++
/*******************************************************//**
The following function generates a hash value for a ulint integer
to a hash table of size table_size, which should be a prime or some
random number to work reliably.
@return	hash value */
UNIV_INLINE
ulint
ut_hash_ulint(
/*==========*/
	ulint	 key,		/*!< in: value to be hashed */
	ulint	 table_size);	/*!< in: hash table size */
```
In case of the modulus being equal to a power of 2, xoring with 1653893711 does not make the distribution better and makes it equivalent to `const ulong pos= key % SIZE;` which is just taking the last 13 bits in this particular case (when SIZE == 8192).

Consider using something like `const ulong pos= (key * <BIG_PRIME>) % SIZE;` where `BIG_PRIME < SIZE`.
BIG_PRIME == 8191, for instance.

Reply:
Good catch - since this calculation is the only way to index into the bitset, OK to make SIZE 8191 in order not to introduce another constant?

Reply:
Yes, making `SIZE=8191` is even better.